pkg/controller/datavolume/snapshot-clone-controller.go

```diff
@@ -0,0 +1,488 @@
+	}
+	sourceNamespace := dataVolume.Spec.Source.Snapshot.Namespace
+	if sourceNamespace == "" {
+		sourceNamespace = dataVolume.Namespace
+	}
+	token, ok := dataVolume.Annotations[cc.AnnCloneToken]
+	if !ok {
+		return errors.Errorf("no clone token")
+	}
+	pvc.Annotations[cc.AnnCloneToken] = token
+	pvc.Annotations[cc.AnnCloneRequest] = sourceNamespace + "/" + dataVolume.Spec.Source.Snapshot.Name
+	return nil
+}
+
+func (r SnapshotCloneReconciler) sync(log logr.Logger, req reconcile.Request) (dataVolumeCloneSyncResult, error) {
+	syncRes, syncErr := r.syncSnapshotClone(log, req)
+	if err := r.syncUpdate(log, &syncRes.dataVolumeSyncResult); err != nil {
+		syncErr = err
+	}
+	return syncRes, syncErr
+}
+
+func (r SnapshotCloneReconciler) syncSnapshotClone(log logr.Logger, req reconcile.Request) (dataVolumeCloneSyncResult, error) {
+	res, syncErr := r.syncCommon(log, req, r.cleanup, r.prepare)
+	syncRes := dataVolumeCloneSyncResult{dataVolumeSyncResult: *res}
+	if syncErr != nil || syncRes.result != nil {
+		return syncRes, syncErr
+	}
+
+	pvc := syncRes.pvc
+	pvcSpec := syncRes.pvcSpec
+	datavolume := syncRes.dvMutated
+	transferName := getTransferName(datavolume)
+
+	pvcPopulated := pvcIsPopulated(pvc, datavolume)
+	_, prePopulated := datavolume.Annotations[cc.AnnPrePopulated]
+
+	if pvcPopulated || prePopulated {
+		return syncRes, nil
+	}
+
+	nn := types.NamespacedName{Namespace: datavolume.Spec.Source.Snapshot.Namespace, Name: datavolume.Spec.Source.Snapshot.Name}
+	snapshot := &snapshotv1.VolumeSnapshot{}
+	if err := r.client.Get(context.TODO(), nn, snapshot); err != nil {
+		return syncRes, err
+	}
+	if snapshot.Status == nil || snapshot.Status.ReadyToUse == nil || !*snapshot.Status.ReadyToUse {
+		r.log.V(3).Info("snapshot not ready to use, can't clone", "namespace", snapshot.Namespace, "name", snapshot.Name)
+		syncRes.result = &reconcile.Result{RequeueAfter: 2 * time.Second}
+		return syncRes, nil
+	}
+
+	fallBackToHostAssisted, err := r.evaluateFallBackToHostAssistedNeeded(datavolume, pvcSpec, snapshot)
+	if err != nil {
+		return syncRes, err
+	}
+
+	if pvc == nil {
+		if !fallBackToHostAssisted {
+			res, err := r.reconcileRestoreSnapshot(log, datavolume, snapshot, pvcSpec, transferName, &syncRes)
+			syncRes.result = &res
+			return syncRes, err
+		}
+
+		if err := r.createTempHostAssistedSourcePvc(datavolume, snapshot, pvcSpec, &syncRes); err != nil {
+			return syncRes, err
+		}
+		targetHostAssistedPvc, err := r.createPvcForDatavolume(datavolume, pvcSpec, r.updateAnnotations)
+		if err != nil {
+			if cc.ErrQuotaExceeded(err) {
+				r.syncDataVolumeStatusPhaseWithEvent(&syncRes, cdiv1.Pending, nil,
+					Event{
+						eventType: corev1.EventTypeWarning,
+						reason:    cc.ErrExceededQuota,
+						message:   err.Error(),
+					})
+			}
+			return syncRes, err
+		}
+		pvc = targetHostAssistedPvc
+	}
+
+	if fallBackToHostAssisted {
+		if err := r.ensureExtendedToken(pvc); err != nil {
+			return syncRes, err
+		}
+	} else {
```

Comment:
drop this else, and `return syncRes, syncErr` just above it

Reply:
done, thx